File: app/src/main/java/com/zknw/unoduo/game/Bot.kt
package com.zknw.unoduo.game

import kotlin.random.Random

enum class Difficulty(val label: String, val botName: String, val blurb: String) {
    EASY("Facile", "Bot facile", "Joue au hasard et oublie souvent de poser."),
    MEDIUM("Moyen", "Bot moyen", "Joue correctement, mais se trompe encore."),
    HARD("Difficile", "Bot difficile", "Ne gâche rien et te bloque dès qu'il peut.")
}

/** What the bot decided to do on its turn. */
sealed interface BotMove {
    data class Play(val cardId: Int, val color: CardColor?) : BotMove
    data object Draw : BotMove
    data object Pass : BotMove
}

/**
 * The solo opponent. It decides from a [GameView] — the very same snapshot a human
 * player receives — so it never sees anyone's hand and cannot cheat, whatever the
 * difficulty. What changes with difficulty is how much thought goes into the choice.
 */
object Bot {

    fun decide(view: GameView, difficulty: Difficulty, rng: Random): BotMove {
        if (!view.yourTurn) return BotMove.Draw
        val legal = view.hand.filter { it.id in view.legal }

        if (legal.isEmpty()) {
            // Nothing playable: either decline the card just drawn, or draw one.
            return if (view.canPass) BotMove.Pass else BotMove.Draw
        }

        // Drawing a card you did not need is the beginner mistake that actually costs
        // games: simply playing *something* every turn is most of what good play is.
        // So this, rather than which card gets picked, is what sets the levels apart.
        if (view.canDraw && rng.nextInt(100) < ditherPercent(difficulty)) {
            return BotMove.Draw
        }

        val card = choose(legal, view, difficulty, rng)
        val color = if (card.isWild) chooseColor(view, card, difficulty, rng) else null
        return BotMove.Play(card.id, color)
    }

    private fun choose(
        legal: List<Card>,
        view: GameView,
        difficulty: Difficulty,
        rng: Random
    ): Card = when (difficulty) {
        // No plan at all, which is exactly what makes it beatable.
        Difficulty.EASY -> legal[rng.nextInt(legal.size)]

        // Mostly sensible, but it slips often enough to be caught out.
        Difficulty.MEDIUM ->
            if (rng.nextInt(100) < MEDIUM_SLIP_PERCENT) legal[rng.nextInt(legal.size)]
            else legal.maxBy { score(it, view, sharp = false) }

        Difficulty.HARD -> legal.maxBy { score(it, view, sharp = true) }
    }

    /**
     * How much the bot wants to get rid of a card.
     *
     * The ranking of the attacking cards comes straight from the house rules, and is
     * not the one you would expect: eating a +2 does **not** cost you your turn, so a
     * +2 thrown at someone holding a single card does not stop them going out. A Skip
     * does, and so does a +4. That is why Skip outranks +2 here when it matters.
     */
    private fun score(card: Card, view: GameView, sharp: Boolean): Int {
        // A sharp bot reacts a turn earlier than a merely sensible one.
        val closest = view.rivals.minOfOrNull { it.cards } ?: 7
        val threatened = closest <= if (sharp) 2 else 1

        // In a duel a Skip — and a Reverse, which is the same thing — hands the turn
        // straight back to you. That is a free card shed every time, not just a way to
        // block someone about to go out, and it is the single biggest edge available.
        val duel = view.playerCount == 2
        var value = when (card.kind) {
            CardKind.NUMBER -> 10
            CardKind.SKIP, CardKind.REVERSE -> when {
                sharp && duel -> 80
                threatened -> 70
                else -> 20
            }
            CardKind.DRAW_TWO -> if (threatened) 40 else 30
            CardKind.WILD -> 5
            CardKind.WILD_DRAW_FOUR -> if (threatened) 65 else 6
        }

        if (sharp) {
            // Playing a card sets the colour, so favouring the colour you hold most of
            // keeps you able to follow on your next turn.
            if (card.color.isRealColor) {
                value += 2 * view.hand.count { it.color == card.color && it.id != card.id }
            }
            // With two cards left, shed the plain one: a wild guarantees the last move.
            if (view.hand.size == 2 && !card.isWild) value += 15
            // Going out wins on the spot, whatever the card is worth in the abstract.
            if (view.hand.size == 1) value += 1000
        }

        return value
    }

    private fun chooseColor(
        view: GameView,
        played: Card,
        difficulty: Difficulty,
        rng: Random
    ): CardColor {
        if (difficulty == Difficulty.EASY) {
            return CardColor.playable[rng.nextInt(CardColor.playable.size)]
        }
        val rest = view.hand.filterNot { it.id == played.id }
        val best = CardColor.playable.maxBy { color ->
            // Count, then a nudge for holding an action card in that colour: it is
            // worth more than a bare number if the colour has to be defended.
            rest.count { it.color == color } * 4 +
                rest.count { it.color == color && it.kind != CardKind.NUMBER }
        }
        // A hand with no coloured card left makes every colour equal; pick at random
        // rather than always answering red.
        return if (rest.none { it.color == best }) {
            CardColor.playable[rng.nextInt(CardColor.playable.size)]
        } else {
            best
        }
    }

    /** How often the bot needlessly draws instead of playing. */
    private fun ditherPercent(difficulty: Difficulty): Int = when (difficulty) {
        Difficulty.EASY -> 25
        Difficulty.MEDIUM -> 8
        Difficulty.HARD -> 0
    }

    private const val MEDIUM_SLIP_PERCENT = 30
}

File: web/src/bot.js
// The solo opponent, ported from the Android app's Bot.
//
// Same caveat as the engine: this is a literal port, kept in step by
// web/test/engine-conformance.mjs, which makes both versions play the same games with
// the same random source and compares every single move. The order of the random draws
// matters as much as the decisions themselves — dithering is rolled before the card is
// chosen, and the colour after — so it is preserved exactly.

import { Color, Kind, PLAYABLE_COLORS, isWild, view } from './engine.js';

export const Difficulty = {
  EASY: 'EASY',
  MEDIUM: 'MEDIUM',
  HARD: 'HARD',
};

export const DIFFICULTY_ORDER = [Difficulty.EASY, Difficulty.MEDIUM, Difficulty.HARD];

export const DIFFICULTY_INFO = {
  EASY: {
    label: 'Facile',
    botName: 'Bot facile',
    blurb: 'Joue au hasard et oublie souvent de poser.',
  },
  MEDIUM: {
    label: 'Moyen',
    botName: 'Bot moyen',
    blurb: 'Joue correctement, mais se trompe encore.',
  },
  HARD: {
    label: 'Difficile',
    botName: 'Bot difficile',
    blurb: 'Ne gâche rien et te bloque dès qu\'il peut.',
  },
};

const MEDIUM_SLIP_PERCENT = 30;

/** How often the bot needlessly draws instead of playing. */
function ditherPercent(difficulty) {
  if (difficulty === Difficulty.EASY) return 25;
  if (difficulty === Difficulty.MEDIUM) return 8;
  return 0;
}

/** Kotlin's maxBy keeps the first element reaching the maximum. */
function firstMaxBy(items, score) {
  let best = items[0];
  let bestScore = score(best);
  for (let i = 1; i < items.length; i++) {
    const s = score(items[i]);
    if (s > bestScore) { best = items[i]; bestScore = s; }
  }
  return best;
}

/**
 * How much the bot wants to get rid of a card.
 *
 * The ranking of the attacking cards comes from the house rules, and is not the one you
 * would expect: eating a +2 does not cost you your turn, so a +2 thrown at someone
 * holding a single card does not stop them going out. A Skip does, and so does a +4.
 */
function score(card, v, sharp) {
  const counts = v.ri.map((r) => r.c);
  const closest = counts.length ? Math.min(...counts) : 7;
  const threatened = closest <= (sharp ? 2 : 1);

  // In a duel a Skip hands the turn straight back: a free card shed every time.
  const duel = view.playerCount(v) === 2;
  let value;
  switch (card.k) {
    case Kind.NUMBER: value = 10; break;
    case Kind.SKIP:
    case Kind.REVERSE:
      value = (sharp && duel) ? 80 : (threatened ? 70 : 20);
      break;
    case Kind.DRAW_TWO: value = threatened ? 40 : 30; break;
    case Kind.WILD: value = 5; break;
    case Kind.DRAW_FOUR: value = threatened ? 65 : 6; break;
    // Same card, twice the bite: never worth less than a +4.
    case Kind.DRAW_EIGHT: value = threatened ? 90 : 8; break;
    // There is one in the whole deck: worth holding, worth more still on somebody who is
    // about to go out.
    case Kind.DRAW_TWELVE: value = threatened ? 120 : 10; break;
    // Fifty cards. There is no situation in which sitting on this beats playing it, and
    // the bot should not pretend otherwise.
    case Kind.DRAW_FIFTY: value = 400; break;
    // Two cards gone for the price of one, and the colour is yours to pick.
    case Kind.DOUBLE_PLAY: value = v.h.length > 2 ? 60 : 4; break;
    // A Joker that also buys information: a little more than a bare one, no more.
    case Kind.SPY: value = 9; break;
    default: value = 0;
  }

  if (sharp) {
    if (card.c !== Color.WILD) {
      value += 2 * v.h.filter((c) => c.c === card.c && c.i !== card.i).length;
    }
    if (v.h.length === 2 && !isWild(card)) value += 15;
    if (v.h.length === 1) value += 1000;
  }
  return value;
}

function chooseColor(v, played, difficulty, rng) {
  if (difficulty === Difficulty.EASY) {
    return PLAYABLE_COLORS[rng.nextIntBelow(PLAYABLE_COLORS.length)];
  }
  const rest = v.h.filter((c) => c.i !== played.i);
  const best = firstMaxBy(PLAYABLE_COLORS, (color) =>
    rest.filter((c) => c.c === color).length * 4 +
    rest.filter((c) => c.c === color && c.k !== Kind.NUMBER).length);
  if (!rest.some((c) => c.c === best)) {
    return PLAYABLE_COLORS[rng.nextIntBelow(PLAYABLE_COLORS.length)];
  }
  return best;
}

/**
 * Decides from the same snapshot a human player receives, so the bot never sees anyone's
 * hand and cannot cheat, whatever the level.
 */
export function decide(v, difficulty, rng) {
  if (!view.yourTurn(v)) return { kind: 'draw' };
  const legal = v.h.filter((c) => v.l.includes(c.i));

  if (!legal.length) {
    return view.canPass(v) ? { kind: 'pass' } : { kind: 'draw' };
  }

  // Drawing a card you did not need is the beginner mistake that actually costs games.
  if (view.canDraw(v) && rng.nextIntBelow(100) < ditherPercent(difficulty)) {
    return { kind: 'draw' };
  }

  // Mid Coup double there is nothing to draw, so the same hesitation comes out as
  // stopping early instead — otherwise the easy bot never wastes a bonus.
  if (view.inBonus(v) && rng.nextIntBelow(100) < ditherPercent(difficulty)) {
    return { kind: 'pass' };
  }

  let card;
  if (difficulty === Difficulty.EASY) {
    card = legal[rng.nextIntBelow(legal.length)];
  } else if (difficulty === Difficulty.MEDIUM) {
    card = rng.nextIntBelow(100) < MEDIUM_SLIP_PERCENT
      ? legal[rng.nextIntBelow(legal.length)]
      : firstMaxBy(legal, (c) => score(c, v, false));
  } else {
    card = firstMaxBy(legal, (c) => score(c, v, true));
  }

  const color = isWild(card) ? chooseColor(v, card, difficulty, rng) : null;
  return { kind: 'play', cardId: card.i, color };
}
